### PR TITLE
Add basic heap stats

### PR DESCRIPTION
**Pull request by Jake-Shadle**
Also adds support for replacing the system allocator with mimalloc.

Closes: #984

### CODE REVIEW

**Comment by markmandel** [comment]
Curious what differences you saw with the switch to `mimalloc` ?

**Comment by Jake-Shadle** [comment]
I've had good experiences with it in the past, I didn't do any extensive testing with it though, just did it because I was changing the global allocate anyways.

File: src/alloc.rs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+mod metrics;
```
Comment (markmandel) [inline src/alloc.rs:1]: Apache licence please.

File: src/alloc/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
+/// Spawns a task to periodically update the heap stat metrics from our tracking allocator
```
Comment (markmandel) [inline src/alloc/metrics.rs:1]: Apache licence please.

File: src/alloc/tracking.rs [hunk 3]
```diff
@@ -0,0 +1,308 @@
+use std::alloc::Layout;
```
Comment (markmandel) [inline src/alloc/tracking.rs:1]: Apache licence please.

File: src/alloc/tracking.rs [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+            .cumul_free_count
+            .fetch_add(stats.cumul_free_count, Relaxed);
+        stats.cumul_free_size += self
+            .cumul_free_size
+            .fetch_add(stats.cumul_free_size, Relaxed);
+
+        stats
+    }
+
+    // pub fn reset(&self) {
```
Comment (markmandel) [inline src/alloc/tracking.rs:257]: Guessing this was left commented out by mistake and not deleted.

File: src/alloc/metrics.rs [hunk 5]
```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2023 Google LLC
```
Comment (markmandel) [inline src/alloc/metrics.rs:2]:
```suggestion
 * Copyright 2024 Google LLC
```

😃 made me question what year we're in.

**Review by XAMPPRocky — APPROVED** [review]

**Pull request merged**